#1.mtcars

mtcars 데이터셋(mtcars.csv)의 qsec 컬럼을 최소 최대 척도(Min-Max Scale)로 변환후 0.5보다 큰값을 가지는 레코드 수를 구하시오

In [1]:
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/HyeonSu-Kang/data/main/mtcars.csv')
df.head(2)

Unnamed: 0   mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  \
0      Mazda RX4  21.0    6  160.0  110   3.9  2.620  16.46   0   1     4   
1  Mazda RX4 Wag  21.0    6  160.0  110   3.9  2.875  17.02   0   1     4   

   carb  
0     4  
1     4

In [2]:
#[1] MinMaxSclaer 사용
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/HyeonSu-Kang/data/main/mtcars.csv')
df.head(2)
from sklearn.preprocessing import MinMaxScaler
df['qsec'] = MinMaxScaler().fit_transform(df[['qsec']])
count = df['qsec'] > 0.5
print(count.sum())

#[2] StandardScaler 사용
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/HyeonSu-Kang/data/main/mtcars.csv')
df.head(2)
from sklearn.preprocessing import StandardScaler
temp = StandardScaler().fit_transform(df[['qsec']])
# 표준정규분포(평균:0,표준편차:1)
print(temp.mean(),temp.std(ddof=0))


9
-1.4710455076283324e-15 0.9999999999999999


In [3]:
#[3] min_max_scaler 직접구현하는 방법
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/HyeonSu-Kang/data/main/mtcars.csv')
df.head(2)
def min_max_scaler(a):
  return(a-a.min())/(a.max()-a.min())
temp = min_max_scaler(df['qsec'])
print(temp.min(),temp.max())

result1 = len(temp[temp>0.5])
result2 = temp[temp>0.5].shape[0]
result3 = (temp>0.5).sum()
print(result1,result2,result3)

# StandardScaler 함수 생성
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/HyeonSu-Kang/data/main/mtcars.csv')

def StandardScaler(a):
  return(a - a.mean())/a.std(ddof=0)

temp = StandardScaler(df['qsec'])
print(temp.mean(),temp.std(ddof=0))

0.0 1.0
9 9 9
-1.4710455076283324e-15 0.9999999999999999


#2.보스턴 범죄 데이터

보스턴 데이터 범죄율 칼럼(CRIM)의 top10 중 10번째 범죄율 값으로 1~10위의 범죄율 값을 변경후, 'AGE'변수의 값이 80 이상인 것에 대한 범죄율 평균을 산출하라

In [4]:
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/HyeonSu-Kang/data/main/boston.csv')
df.head(2)

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   

   PTRATIO      B  LSTAT  PRICE  
0     15.3  396.9   4.98   24.0  
1     17.8  396.9   9.14   21.6

In [5]:
#[1] 보스턴 데이터 범죄율 컬럼('CRIM')의 top10 구하기
A = df['CRIM'].sort_values(ascending=False).head(10)
#[2] 10번째 범죄율 값으로 1~10위의 범죄율 값을 변경하기
df.loc[A.index,'CRIM'] = A.iloc[-1]
#[3] (전체 데이터에서)'AGE' 변수의 값이 80 이상인 것에 대한 범죄율 평균 구하기
result1 = df.loc[df['AGE'] >= 80,'CRIM'].mean()
result2 = df.loc[df.AGE >=80,'CRIM'].mean()
print(result1,result2)

5.759386625 5.759386625


In [6]:
#[2] DataFrame.sort_values 사용하기
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/HyeonSu-Kang/data/main/boston.csv')
df.head(2)

A = df.sort_values('CRIM',ascending=False)
df.loc[A.index,'CRIM'] = A.iloc[-1,0]
A =df.loc[df.AGE>=80,'CRIM'].mean()

#3.하우징 데이터

###3.1하우징 데이터

* 주어진 데이터의 첫번째 행 부터 순서대로 80%까지의 데이터를 추출
* total_bedrooms 변수의 결측값(NA)을 'total_bedrooms 변수의 중앙값으로 대체하고,
* 대체 전의 total_bedrooms 변수 표준편차과 대체 후의 'total_bedrooms'변수 표준편차 값을 산출하여
* 대체 전 표준편차 값,대체 후 표준편차 값을 구해 출력한다.

In [7]:
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/HyeonSu-Kang/data/main/housing.csv')
print(df.shape)
df.head(2)

(20640, 10)


longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY  
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY

In [13]:
#[1] 계산 사용
#!) 첫번째 행 부터 순서대로 80% 까지의 데이터를 추출
p = int(df.shape[0]*0.8)
#p = round(df.shape[0]*0.8)
A = df.loc[:p]

#2) total_bedrooms 변수의 결측값(NA)를 total_bedrooms 변수의 중앙값으로 대체
A_before = A['total_bedrooms']
A_after = A['total_bedrooms'].fillna(A['total_bedrooms'].median())

#3) 대체 전 표본편차 값 - 대채 후 표분편차 값
result1 = A_before.std(ddof=0) - A_after.std(ddof=0)
result2 = A_before.std(ddof=1) - A_after.std(ddof=1)
print(f'{result1:.4f},{result2:.4f}')
print(round(result1),round(result2))

1.9748,1.9750
2 2


###3.2 하우징 데이터

* 하우징 데이터에서 'latitude' 칼럼의 이상치를 찾아 이상치들의 합을 산출하시오
* 이상치 기준 : 평군 - (표준편차*1.5), 평균 + (표준편차*1.5)



In [32]:
#[0] 데이터준비
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/HyeonSu-Kang/data/main/housing.csv')
df.head(2)

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY  
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY

In [33]:
# 이상치 합 구하기
temp = df['latitude']
t_mean = temp.mean()
t_std = temp.std(ddof=0)

lower = t_mean-1.5*t_std
upper = t_mean+1.5*t_std
a = temp[temp>upper].sum()
b = temp[temp<lower].sum()
print(a+b)

45815.75


In [61]:
# 이상치에 해당하는 데이터 개수
df = pd.read_csv('https://raw.githubusercontent.com/HyeonSu-Kang/data/main/housing.csv')
temp = df['latitude']
mean,std = temp.mean(),temp.std(ddof=0)
lower = mean - 1.5*std
upper = mean + 1.5*std
a = temp[temp<lower].shape[0]
b = temp[temp>upper].shape[0]
result = a+b
print(result)

cond = (temp < lower) | (temp > upper)
result2 = temp[cond].shape[0]
print(result2)

# 이상치에 해당하는 데이터 검색
cond = (temp<lower) | (temp>upper)
df[cond]

1151
1151


longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
1053     -121.83     39.76                12.0       9831.0          1921.0   
1054     -121.82     39.76                23.0       6010.0          1116.0   
1055     -121.86     39.78                12.0       7653.0          1578.0   
1056     -121.85     39.77                17.0       5273.0          1177.0   
1057     -121.84     39.76                14.0       2351.0           620.0   
...          ...       ...                 ...          ...             ...   
20635    -121.09     39.48                25.0       1665.0           374.0   
20636    -121.21     39.49                18.0        697.0           150.0   
20637    -121.22     39.43                17.0       2254.0           485.0   
20638    -121.32     39.43                18.0       1860.0           409.0   
20639    -121.24     39.37                16.0       2785.0           616.0   

       population  households  median_income  median_house_value  \
1053       4644.0      1775.0         3.1142            112600.0   
1054       2710.0      1149.0         3.0068            107300.0   
1055       3628.0      1494.0         3.0905            117800.0   
1056       2446.0      1199.0         1.9362             89900.0   
1057       1215.0       548.0         2.3155            102300.0   
...           ...         ...            ...                 ...   
20635       845.0       330.0         1.5603             78100.0   
20636       356.0       114.0         2.5568             77100.0   
20637      1007.0       433.0         1.7000             92300.0   
20638       741.0       349.0         1.8672             84700.0   
20639      1387.0       530.0         2.3886             89400.0   

      ocean_proximity  
1053           INLAND  
1054           INLAND  
1055           INLAND  
1056           INLAND  
1057           INLAND  
...               ...  
20635          INLAND  
20636          INLAND  
20637          INLAND  
20638          INLAND  
20639          INLAND  

[1151 rows x 10 columns]

#4.캘리포니어 집값 데이터

* 결측치를 포함하는 모든 행을 제거한 후, 처음부터 순서대로 70%를 추출하여, 'housing_median_age' 컬럼의 사분위수 Q1의 값을 구하시오

* 정수형으로
* 한 문자형 컬럼에 빈칸''인 데이터 1건이 있어서
* 빈칸을 결측으로 처리시 빈칸 제거전 242건에서 빈칸 제거후 243건의 행이 제거됨
* 하지만 제거하지 않고 1분위수를 구하더라도 정답은 같음

In [87]:
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/HyeonSu-Kang/data/refs/heads/main/housing03.csv')
df.head(2)

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY  
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY

In [93]:
# 1) 결측치 확인
print(df.isna().sum().sum()) #207

# 2) 결측치를 포함한 모든 행 제거
df = df.dropna(how='any',axis=0)

0


In [94]:
#[3] 처음부터 순서대로 70%를 추출
from sklearn.model_selection import train_test_split
A,_ = train_test_split(df,test_size=0.3,shuffle=False)
#다른방법 df.iloc[:int(len(df)*0.7)]

#[4] 'housing_median_age' 칼럼의 사분위수 Q1의 값 구하기
result = df['housing_median_age'].quantile(0.25)

#[5] 정수로 출력하기
print(int(result))


18


##5.연도별 나라별 유병률 데이터


* 2000년 데이터에 대해 국가 전체 유병률 평균 값보다 큰  값을 가진 국가 개수는?
(year 변수와 국가들의 결핵 유병률 값을 가진 데이터셋이 주어짐)

In [118]:
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/HyeonSu-Kang/data/refs/heads/main/worlddata.csv')
df.head()

year  Afghanistan  Albania  Algeria  Andorra  Angola  Antigua & Barbuda  \
0  1999            0     89.0     25.0    245.0   217.0              102.0   
1  2000            0    132.0      0.0    138.0    57.0              128.0   
2  2001            0     54.0     14.0    312.0    45.0               45.0   
3  2002            0      4.9      0.7     12.4     5.9                4.9   

   Argentina  Armenia  Australia  ...  Tanzania    USA  Uruguay  Uzbekistan  \
0      193.0     21.0      261.0  ...      36.0  249.0    115.0        25.0   
1       25.0    179.0       72.0  ...       6.0  158.0     35.0       101.0   
2      221.0     11.0      212.0  ...       1.0   84.0    220.0         8.0   
3        8.3      3.8       10.4  ...       5.7    8.7      6.6         2.4   

   Vanuatu  Venezuela  Vietnam  Yemen  Zambia  Zimbabwe  
0     21.0      333.0      111    6.0    32.0      64.0  
1     18.0      100.0        2    0.0    19.0      18.0  
2     11.0        3.0        1    0.0     4.0       4.0  
3      0.9        7.7        2    0.1     2.5       4.7  

[4 rows x 194 columns]

In [116]:
mean = (df['year']==2000).mean()
(df[df['year'] == 2000] > mean).shape[0]

1

In [132]:
df2000 = df[df['year']==2000].T
data = df2000.iloc[1:,0]
mean = data.mean()
data[data>mean].shape[0]

76

##6.타이타닉 데이터


* 각 열의 결측치 비율을 확인 한 후, 결측치의 비율이 가장 높은 변수명을 구하시오




In [134]:
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/HyeonSu-Kang/data/refs/heads/main/titanic_train03.csv')
df.head(2)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   

   Parch     Ticket     Fare Cabin Embarked  
0      0  A/5 21171   7.2500   NaN        S  
1      0   PC 17599  71.2833   C85        C

In [153]:
non = df.isna().sum()

result1 = non.sort_values(ascending=False).index[0]
result2 = non.index[non.argmax()]
print(result1,result2)

Age Age
